File: ABS/049C.cpp
#include <iostream>
#include <string>
#include <algorithm>
using namespace std;

string divide[4] = {"dream", "dreamer", "erase", "eraser"};

int main() {
    string S;
    cin >> S;

    reverse(S.begin(), S.end());
    for (int i=0; i<4; ++i) reverse(divide[i].begin(), divide[i].end());

    bool can = true;
    int len = S.size();
    for (int i=0; i<len;) {
        bool can2 = false;
        for (int j=0; j<4; ++j) {
            if (S.substr(i, divide[j].size()) == divide[j]) {
                can2 = true;
                i += divide[j].size();
            }
        }
        if (!can2) {
            can = false;
            break;
        }
    }

    if (can) cout << "YES" << endl;
    else cout << "NO" << endl;

    return 0;
}

File: ABS/049C.cs
using System;
class Program
{
    static void Main(string[] args) {
        string S = Console.ReadLine();
        int len = S.Length;
        int pos = 0;
        int flag = 0;
        while (pos < len) {
            if (pos+7 <= len && S.Substring(pos, 7)=="dreamer") {
                pos += 7;
                flag = 2;
            }
            else if (pos+6 <= len && S.Substring(pos, 6)=="eraser") {
                pos += 6;
                flag = 1;
            }
            else if (pos+5 <= len && S.Substring(pos, 5)=="dream") {
                pos += 5;
                flag = 0;
            }
            else if (pos+5 <= len && S.Substring(pos, 5)=="erase") {
                pos += 5;
                flag = 0;
            }
            else if (flag > 0) {
                pos -= flag;
                flag = 0;
            }
            else break;
        }
        if (pos==len) Console.WriteLine("YES");
        else Console.WriteLine("NO");

        return;
    }
}